infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java

```diff
@@ -421,12 +421,35 @@ public synchronized void tokenRequest(CorfuPayloadMsg<TokenRequest> msg,
                 handleTxToken(msg, ctx, r);
                 return;
 
+            case TokenRequest.TK_QUERY_ALL:
+                handleMultiStreamQuery(msg, ctx, r);
+                return;
+
             default:
                 handleAllocation(msg, ctx, r);
                 return;
         }
     }
 
+
+    /**
+     *
+     * this method is able to query multiple stream tails within the same request.
+     * For now it returns the whole stream tails map, but later we will change it
+     * to service a subset of that map.
+     *
+     * @param msg corfu message containing token query
+     * @param ctx netty ChannelHandlerContext
+     * @param r   server router
+     */
+    public void handleMultiStreamQuery(CorfuPayloadMsg<TokenRequest> msg,
+                                      ChannelHandlerContext ctx, IServerRouter r) {
+        Token token = new Token(getGlobalLogTail().get() - 1, bootstrapEpoch);
+        r.sendResponse(ctx, msg, CorfuMsgType.TOKEN_RES.payloadMsg(new TokenResponse(
+                TokenType.NORMAL, TokenResponse.NO_CONFLICT_KEY, token, Collections.emptyMap(), streamTails)));
```

Comment:
a copy of streamTails should be made before sendResponse is called, otherwise the map can change before its pushed on the wire.